### PR TITLE
Offline update operations 4078

### LINKED ISSUE #4078: Update operations (e.g. marking email unread) fail with error offline because we can't encrypt entities
When we want to re-encrypt an instance we need to get it's previous key.
If it was already updated with symmetric key then it is stored as a field: `_ownerEncSessionKey`.
This key is encrypted with a group key. However we don't have group keys offline (to get group key stored on `GroupMembership` we need `userKey` and to get that we need `passPhraseKey` which is encrypted with the session).

session -> password -> userKey -> groupKey -> ownerEncSessionKey

From the user perspective there should be no difference for offline and online login and it should behave the same: produce an offline message.

## Implementation hint
We normally encrypt only in two places: `ServiceExecutor` and `EntityRestClient`. We should add some handling for offline there to not try to encrypt and just bail out with offline error.

Related to #3888

**Test notes**
Log in while offline and verify, that all actions result in the expected behavior:

- [x] send mail
- [x] reply or forward 
- [x]  mark read
- [x] move mail 
- [x] delete mail 
- [x] archive mail 
- [x] report 
- [x] show email headers (from top menu or shortcut)
- [ ] Add spam rule
- [ ]  add inbox rule 
- [ ] add folder 
- [ ] export

### PR DESCRIPTION
**Pull request by sarashub**
This PR introduces LoginIncompleteError that is thrown when update operations fail due to offline (and therefore incomplete) login (see individual commit messages). Close #4078

### CODE REVIEW

File: src/api/worker/facades/UserFacade.ts [hunk 1]
```diff
@@ -83,7 +84,17 @@ export class UserFacade implements AuthHeadersProvider {
...
 	getUserGroupKey(): Aes128Key {
 		// the userGroupKey is always written after the login to this.groupKeys
-		return assertNotNull(this.groupKeys.get(this.getUserGroupId()), "User is not logged in")
+		//if the user has only logged in offline this has not happened
+		const userGroupKey = this.groupKeys.get(this.getUserGroupId())
+		if (userGroupKey == null) {
+			if (this.isPartiallyLoggedIn()) {
+				throw new LoginIncompleteError("userGroupKey not available")
+			} else {
+				throw new Error("User is not logged in")
```
Comment (ganthern) [inline src/api/worker/facades/UserFacade.ts:93]: `getUserGroupKey` previously asserted not null, why is `userGroupKey === null` possible now even when not partially logged in? is this a case of an "unreachable" situation?

**Review by ganthern — COMMENTED** [review]
* [x] smoke tested
* [x] good commit messages
* [x] doc comments where necessary
* [x] commits are squashed appropriately
* [x] sufficient test notes
* [x] changes reviewed
* [ ] ~translation keys are in Phrase~

the commit messages should reference the issue id in the last line. one of them can use the `fix` or `close` keywords so the PR gets automatically linked to the issue.

Edit: I think this needs to be rebased so the tests pass.

File: src/api/worker/facades/UserFacade.ts (hunk 1 above)
Comment (charlag) [reply src/api/worker/facades/UserFacade.ts:93]: It shouldn't be reached, it's equivalent of what would happen before. The only way to be fully logged in is to unlock the key.

Comment (ganthern) [reply src/api/worker/facades/UserFacade.ts:93]: then I'd propose to make this a ProgrammingError and put that info in the message

**Review by ganthern — APPROVED** [review]

**Pull request merged**